quesma/optimize/shorten_time_range.go

```diff
@@ -0,0 +1,248 @@
+package optimize
+
+import (
+	"fmt"
+	"quesma/logger"
+	"quesma/model"
+	"strconv"
+)
+
+type shortenTimeRange struct{}
+
+type timeRange struct {
+	columnName string
+	lowerLimit int64
+	upperLimit int64
+	direction  model.OrderByDirection
+}
+
+func (s shortenTimeRange) validateSelectedColumns(columns []model.Expr) bool {
+	// The main purpose is to disallow window functions for which this optimization might be hard to reason about
+	// (and could be invalid). The allowed Expr types are whitelisted here rather than blacklisted (window functions)
+	// to be less error-prone in the future.
+	for i, column := range columns {
+		if _, ok := column.(model.ColumnRef); ok {
+			continue
+		}
+		if _, ok := column.(model.LiteralExpr); ok {
+			continue
+		}
+		if aliasedExpr, ok := column.(model.AliasedExpr); ok && s.validateSelectedColumns([]model.Expr{aliasedExpr.Expr}) {
+			continue
+		}
+		if functionExpr, ok := column.(model.FunctionExpr); ok && s.validateSelectedColumns(functionExpr.Args) {
+			continue
+		}
+
+		logger.Debug().Msgf("Query not eligible for time range optimization: column at index %d is an unsupported %T", i, column)
+		return false
+	}
+	return true
+}
+
+func (s shortenTimeRange) findOrderByColumn(selectCommand *model.SelectCommand) (string, model.OrderByDirection, bool) {
+	if len(selectCommand.OrderBy) != 1 {
+		logger.Debug().Msg("Query not eligible for time range optimization: ORDER BY longer than 1")
+		return "", model.DefaultOrder, false
+	}
+
+	if orderByColumn, ok := selectCommand.OrderBy[0].Expr.(model.ColumnRef); ok {
+		return orderByColumn.ColumnName, selectCommand.OrderBy[0].Direction, true
+	}
+
+	logger.Debug().Msg("Query not eligible for time range optimization: ORDER BY not a column reference")
+	return "", model.DefaultOrder, false
+}
+
+func (s shortenTimeRange) findTimeLimits(selectCommand *model.SelectCommand, orderByColumnName string) (*int64, *int64) {
+	var lowerLimit, upperLimit *int64
+
+	visitor := model.NewBaseVisitor()
+	visitor.OverrideVisitInfix = func(b *model.BaseExprVisitor, e model.InfixExpr) interface{} {
+		if columnName, ok := e.Left.(model.ColumnRef); ok && columnName.ColumnName == orderByColumnName {
+			switch e.Op {
+			case "<", "<=":
+				if functionExpr, ok := e.Right.(model.FunctionExpr); ok && functionExpr.Name == "fromUnixTimestamp64Milli" {
+					upperBoundValue := functionExpr.Args[0].(model.LiteralExpr).Value.(int64)
+					upperLimit = &upperBoundValue
+				}
+			case ">", ">=":
+				if functionExpr, ok := e.Right.(model.FunctionExpr); ok && functionExpr.Name == "fromUnixTimestamp64Milli" {
+					lowerBoundValue := functionExpr.Args[0].(model.LiteralExpr).Value.(int64)
+					lowerLimit = &lowerBoundValue
+				}
+			}
+		}
+		e.Left.Accept(b)
+		e.Right.Accept(b)
+		return e
+	}
+	selectCommand.Accept(visitor)
+
+	return lowerLimit, upperLimit
+}
+
+func (s shortenTimeRange) findTimeRange(selectCommand *model.SelectCommand) *timeRange {
```

Comment:
Different names, check if possible to optimize and find time range.

Reply:
Done